File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:

        length = len(s)
        l = 0
        result = 0
        seen = defaultdict(int)

        for r in range(len(s)):
            seen[s[r]] = 1 + seen.get(s[r], 0)  

            while ((r-l + 1) - max(seen.values())) > k :
                seen[s[l]] -= 1
                l += 1

            result = max(result, r-l+1)
        
        return result

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
    int result = 0;
public:
    int diameterOfBinaryTree(TreeNode* root) {
        subTreeDepth(root);
        return result;
    }
private:
    int subTreeDepth(TreeNode* node) {
        if (!node) return 0;
        int leftDepth = subTreeDepth(node->left);
        int rightDepth = subTreeDepth(node->right);
        result = max(result, leftDepth+rightDepth);
        return 1 + max(leftDepth, rightDepth);
    }
};
